### 4. Với mỗi cột, các giá trị được phân bố như thế nào?


#### Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá trị được phân bố như thế nào?

Ở dữ liệu này, hiện có 28 cột thuộc nhóm numeric:
|   |   |
| - | - |
| 1. year | 15. crude_death_rate|
| 2. infant_mortality| 16. net_migration|
| 3. infant_mortality_male| 17. rate_natural_increase|
| 4. infant_mortality_female | 18. growth_rate|
| 5. life_expectancy| 19. fertility_rate_15_19|
| 6. life_expectancy_male| 20. fertility_rate_20_24|
| 7. life_expectancy_female| 21. fertility_rate_25_29|
| 8. mortality_rate_under5| 22. fertility_rate_30_34|
| 9. mortality_rate_under5_male| 23. fertility_rate_35_39|
| 10. mortality_rate_under5_female| 24. fertility_rate_40_44|
| 11. mortality_rate_1to4| 25. fertility_rate_45_49|
| 12. mortality_rate_1to4_male| 26. total_fertility_rate|
| 13. mortality_rate_1to4_female| 27. gross_reproduction_rate|
| 14. crude_birth_rate| 28. sex_ratio_at_birth|

In [12]:
pd.set_option('display.max_colwidth', 200)
numerical_cols = df[df.select_dtypes(exclude=['object']).columns]
numerical_cols.head(5)

year  infant_mortality  infant_mortality_male  infant_mortality_female  \
0  2036              3.39                   3.76                     3.00   
1  2022              3.76                   4.22                     3.27   
2  2023              3.73                   4.18                     3.25   
3  2024              3.70                   4.14                     3.22   
4  2025              3.67                   4.10                     3.20   

   life_expectancy  life_expectancy_male  life_expectancy_female  \
0            80.90                 77.51                   84.52   
1            79.11                 75.58                   82.89   
2            79.26                 75.73                   83.02   
3            79.40                 75.89                   83.15   
4            79.55                 76.04                   83.29   

   mortality_rate_under5  mortality_rate_under5_male  \
0                   3.93                        4.39   
1                   4.43                        5.02   
2                   4.39                        4.97   
3                   4.35                        4.91   
4                   4.31                        4.86   

   mortality_rate_under5_female  mortality_rate_1to4  \
0                          3.44                 0.54   
1                          3.81                 0.68   
2                          3.78                 0.67   
3                          3.74                 0.65   
4                          3.71                 0.64   

   mortality_rate_1to4_male  mortality_rate_1to4_female  crude_birth_rate  \
0                      0.63                        0.44              7.59   
1                      0.80                        0.54              7.52   
2                      0.79                        0.53              7.43   
3                      0.77                        0.52              7.35   
4                      0.76                        0.51              7.30   

   crude_death_rate  net_migration  rate_natural_increase  growth_rate  \
0             14.65           0.23                 -0.706       -0.682   
1             12.28           0.32                 -0.476       -0.444   
2             12.41           0.31                 -0.498       -0.467   
3             12.55           0.31                 -0.520       -0.489   
4             12.68           0.30                 -0.538       -0.508   

   fertility_rate_15_19  fertility_rate_20_24  fertility_rate_25_29  \
0                   8.5                  60.2                 112.0   
1                   7.5                  56.4                 106.1   
2                   7.6                  56.5                 106.6   
3                   7.7                  56.8                 107.0   
4                   7.7                  56.9                 107.4   

   fertility_rate_30_34  fertility_rate_35_39  fertility_rate_40_44  \
0                  86.3                  32.0                   8.8   
1                  75.5                  26.7                   6.4   
2                  76.2                  27.2                   6.6   
3                  77.1                  27.5                   6.7   
4                  78.0                  27.7                   6.9   

   fertility_rate_45_49  total_fertility_rate  gross_reproduction_rate  \
0                   3.7                1.5555                   0.7529   
1                   2.3                1.4110                   0.6829   
2                   2.4                1.4213                   0.6879   
3                   2.5                1.4316                   0.6929   
4                   2.6                1.4419                   0.6979   

   sex_ratio_at_birth  
0              1.0660  
1              1.0661  
2              1.0660  
3              1.0660  
4              1.0661

Với mỗi cột numeric ta sẽ tính tỉ lệ % giá trị thiếu, min, max. Và lưu kết quả vào Dataframe `nume_col_profiles_df`.
Dataframe này gồm có:
- 3 dòng là `missing_ratio`, `min`, `max` lần lượt là tỉ lệ các giá trị thiếu mỗi cột, min và max.
- 28 cột là các thuộc tính thuộc kiểu dữ liệu **numerical**.

In [14]:
def missing_ratio(c):
    return  c.isna().sum()/len(df) * 100

def min(c):
    return c.min()

def max(c):
    return c.max()

nume_col_profiles_df = numerical_cols.agg([missing_ratio, min, max])
nume_col_profiles_df

year  infant_mortality  infant_mortality_male  \
missing_ratio     0.0              0.00                   0.00   
min            1950.0              1.15                   0.22   
max            2050.0            298.30                 314.65   

               infant_mortality_female  life_expectancy  life_expectancy_male  \
missing_ratio                     0.00             0.00                  0.00   
min                               0.73            10.01                  9.82   
max                             281.22            91.58                 88.69   

               life_expectancy_female  mortality_rate_under5  \
missing_ratio                    0.00                   0.00   
min                             10.21                   1.44   
max                             94.97                 455.12   

               mortality_rate_under5_male  mortality_rate_under5_female  \
missing_ratio                        0.00                          0.00   
min                                  1.66                          1.21   
max                                466.23                        443.65   

               mortality_rate_1to4  mortality_rate_1to4_male  \
missing_ratio                 0.00                      0.00   
min                           0.20                      0.21   
max                         335.59                    342.79   

               mortality_rate_1to4_female  crude_birth_rate  crude_death_rate  \
missing_ratio                        0.00              0.00              0.00   
min                                  0.00              3.62              1.53   
max                                328.31             58.74             97.08   

               net_migration  rate_natural_increase  growth_rate  \
missing_ratio           0.00                  0.000        0.000   
min                  -831.81                 -6.383     -168.944   
max                  1693.01                  4.036      168.887   

               fertility_rate_15_19  fertility_rate_20_24  \
missing_ratio                   0.0                   0.0   
min                             0.0                   9.5   
max                           237.4                 363.3   

               fertility_rate_25_29  fertility_rate_30_34  \
missing_ratio                   0.0                   0.0   
min                            34.1                  23.1   
max                           369.2                 338.3   

               fertility_rate_35_39  fertility_rate_40_44  \
missing_ratio                   0.0                   0.0   
min                             7.2                   0.0   
max                           282.9                 201.3   

               fertility_rate_45_49  total_fertility_rate  \
missing_ratio                   0.0                0.0000   
min                             0.0                0.6621   
max                           105.8                8.0700   

               gross_reproduction_rate  sex_ratio_at_birth  
missing_ratio                   0.0000              0.0000  
min                            -9.0000             -9.0000  
max                             3.9754              1.4696

Nhận thấy không có giá trị nào bị thiếu ở tất cả các cột (phần trăm thiếu bằng 0%).
Với dữ liệu ở cột  `year`, dữ liệu được thu thập và dự đoán từ năm 1950 đến năm 2050.

#### Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào?

Ở dữ liệu này, hiện có 2 cột thuộc nhóm categorical:
1. country_code
2. country_name


In [15]:
pd.set_option('display.max_colwidth', 200)
object_cols = df[df.select_dtypes(include=['object']).columns]
object_cols.head(5)

country_code country_name
0           SI     Slovenia
1           SI     Slovenia
2           SI     Slovenia
3           SI     Slovenia
4           SI     Slovenia

Với mỗi cột categorical, ta tính tỉ lệ % giá trị thiếu, số lượng giá trị khác nhau (không xét giá trị thiếu), list/array các giá trị khác nhau (không xét giá trị thiếu).
- Kết quả được lưu vào dataframe `cate_col_profiles_df`.
- Dataframe này có 3 dòng là `missing_ratio`, `num_diff_vals`, `ratio_diff_vals`; và có 2 cột là các thuộc tính có kiểu dữ liệu **categorical**.

In [16]:
def num_diff_vals(col):
    return len(col.value_counts())

def ratio_diff_vals(col):
    return np.round(col.value_counts(normalize=True),2).to_dict()

cate_col_profiles_df=object_cols.agg([missing_ratio,num_diff_vals,ratio_diff_vals])
cate_col_profiles_df

country_code  \
missing_ratio                                                                                                                                                                                                        0.0   
num_diff_vals                                                                                                                                                                                                        228   
ratio_diff_vals  {'DJ': 0.01, 'BT': 0.01, 'NI': 0.01, 'GV': 0.01, 'CB': 0.01, 'CD': 0.01, 'GB': 0.01, 'UG': 0.01, 'GH': 0.01, 'WI': 0.01, 'AO': 0.01, 'LE': 0.01, 'BR': 0.01, 'NP': 0.01, 'HA': 0.01, 'NU': 0.01, 'PA...   

                                                                                                                                                                                                            country_name  
missing_ratio                                                                                                                                                                                                        0.0  
num_diff_vals                                                                                                                                                                                                        228  
ratio_diff_vals  {'Djibouti': 0.01, 'Bhutan': 0.01, 'Nigeria': 0.01, 'Guinea': 0.01, 'Cambodia': 0.01, 'Chad': 0.01, 'Gabon': 0.01, 'Uganda': 0.01, 'Ghana': 0.01, 'Angola': 0.01, 'Brazil': 0.01, 'Lebanon': 0.01, '...

- Nhận thấy không có giá trị nào bị thiếu.
- Tổng cộng có 228 giá trị khác nhau ở `country_code` và  `country_name`, đại diện cho 228 quốc gia trên Thế Giới